### ISSUE TITLE
Type-unsafe constructors

### ISSUE THREAD

**Issue by matbesancon** [JuliaGraphs/StaticGraphs.jl#5]
I'm not a fan of these constructors (same for StaticDigraph)
```
function StaticGraph(n_v, ss::AbstractVector, ds::AbstractVector)
```

They don't specify an element type for ss and ds, and then need to use `convert(Vector{T},ds)` with T as the smallest type to contain them:
```
function mintype(v::AbstractVector)
    validtypes = [UInt8, UInt16, UInt32, UInt64, UInt128]
    l = v[end]
    for T in validtypes
        l <= typemax(T) && return T
    end
    return eltype(v)
end
```
Shouldn't the input vectors directly specify a type?

**Comment by sbromberger** [comment]
This is actually a really useful constructor for my work :) Basically, when you have a large graph that's saved to disk as an edgelist (think the SNAP datasets, for instance), you can read in the graph using `readdlm` or whatever, and then manipulate the vectors. When you then feed it to the graph constructor, it will automatically optimize the type of the graph so you're not taking more memory than you actually need (that is, if you only have 60k vertices, then your graph can use `UInt16`; if you have fewer than 4 million vertices, then you can fit everything you need into a `UInt32`, etc.)

What specifically don't you like about it, and are there any alternatives that ensure an optimal data structure?

**Comment by mbesancon** [comment]
Oh ok I understand now! 
Shouldn't we still indicate that `ss::AbstractVector, ds::AbstractVector` are vectors containing numbers? That's the only puzzling thing for me, we don't know the content of the vectors from the signature.

**Comment by sbromberger** [comment]
Oh, I guess we could type parameterize them: `ss::AbstractVector{T}, ds::AbstractVector{U} ... where T <: Integer where U <: Integer`?

**Comment by sbromberger** [comment]
So, here's the thing: you can't add vertices to a static graph, so it only makes sense to use the most efficient int type, right? Why would you want to use anything wider than what you need?

**Comment by mbesancon** [comment]
this makes sense, I was more thinking of the cost of looping to find the mintype which can be avoided.
In cases when you call the constructor many times, you might prefer a lightweight version. Also does this not hinder type-stability? We're determining the return type at runtime

**Comment by sbromberger** [comment]
I'm hoping you're not calling this constructor many times :) StaticGraphs are really designed for very large graphs. Once you've constructed it via the source/dest vectors, it's better to save it off using the persistence methods.

If you're using this package for multiple smaller graphs, it's not going to give you better performance than the native SimpleGraph structure.

**Comment by mbesancon** [comment]
I understand, let's keep it simple you're right. I'll change the signature to `ss::AbstractVector{T}, ds::AbstractVector{U} ... where T <: Integer where U <: Integer` in the cleanup PR

**Comment by matbesancon** [comment]
Fixed in #4

**Comment by sbromberger** [comment]
Closed via #4
